### PR TITLE
Create portfolio entity

### PR DESCRIPTION
**Pull request by Cebix90**
Created portfolio entity with tests

### CODE REVIEW

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 1]
```diff
@@ -63,52 +65,56 @@ public void deleteUser(String customUserEmail) {
...
         try (Session session = sessionFactory.openSession()) {
             Transaction transaction = session.beginTransaction();
             CustomUser customUserToUpdate = findUserByEmail(email);
 
             if (customUserToUpdate != null) {
-                validator.accept(value);
+                if (!validator.test(value)) {
+                    throw new IllegalArgumentException(getValidationErrorMessage(fieldName, value));
+                }
                 fieldUpdater.accept(customUserToUpdate);
```
Comment (Arturnix) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:88]: This code block can be one method.

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAO.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    private <T> Portfolio updatePortfolioField(long customUserId, Consumer<Portfolio> fieldUpdater, Predicate<T> validator, T value, String fieldName) {
+        try (Session session = sessionFactory.openSession()) {
+            Transaction transaction = session.beginTransaction();
+            Portfolio portfolioToUpdate = findPortfolioByUserId(customUserId);
+
+            if (portfolioToUpdate != null) {
+                if (!validator.test(value)) {
+                    throw new IllegalArgumentException(getValidationErrorMessage(fieldName, value));
+                }
+                fieldUpdater.accept(portfolioToUpdate);
```
Comment (Arturnix) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAO.java:93]: This code block can be one method with argument _portfolioToUpdate_ which is being returned.

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java (hunk 1 above)
Comment (Arturnix) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:88]: This code block can be one method which takes an argument customUserToUpdate and which is being returned.

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAO.java (hunk 2 above)
Comment (Arturnix) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAO.java:93]: This code block can be one method with argument portfolioToUpdate which is being returned.

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAOTests.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    }
+
+    @Nested
+    class UpdatePortfolio {
+
+        @Nested
+        class UpdatePortfolioStocksCollection {
+            @Test
+            public void testUpdatePortfolioStocksCollection_whenPortfolioExists_thenCorrect() {
+                CustomUser userForTests = createAndSaveUserForTests();
```
Comment (Arturnix) [inline src/test/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAOTests.java:111]: This code repeats, Maybe @BeforeClass or turning it into method will help.

**Review by Arturnix — CHANGES_REQUESTED** [review]
- Adding brief description of this application will help with using it.
- Despite I picked date in the future I didn't get error message: "You can only select a date at least one day in the past." (class FutureDateException, getMessage())
 but this one: "An unexpected error occurred. Please try again."
- Register and login links are visible in home page. Will they be used?
- 3 tests are failing:
MarketStatusAPIHandlerTests -> testGetMarketStatusData_ifCorrectResponse()
TechnicalIndicatorsAPIHandlerTests -> TestEMAData -> testGetEMAData_ifSuccess()
TechnicalIndicatorsAPIHandlerTests -> TestSMADData -> testGetSMADData_ifSuccess()

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java (hunk 1 above)
Comment (Cebix90) [reply src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:88]: I specially created this method this way to update the methods for each important field separately to get better control over them

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAO.java (hunk 2 above)
Comment (Cebix90) [reply src/main/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAO.java:93]: I specially created this method this way to update the methods for each important field separately to get better control over them

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAOTests.java (hunk 3 above)
Comment (Cebix90) [reply src/test/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAOTests.java:111]: Very good idea, I moved initialization of CustomUser object to @BeforeEach

**Comment by Cebix90** [comment]
> * Adding brief description of this application will help with using it.
> * Despite I picked date in the future I didn't get error message: "You can only select a date at least one day in the past." (class FutureDateException, getMessage())
>   but this one: "An unexpected error occurred. Please try again."
> * Register and login links are visible in home page but without routing in controller. Will they be used?
> * 3 tests are failing:
>   MarketStatusAPIHandlerTests -> testGetMarketStatusData_ifCorrectResponse()
>   TechnicalIndicatorsAPIHandlerTests -> TestEMAData -> testGetEMAData_ifSuccess()
>   TechnicalIndicatorsAPIHandlerTests -> TestSMADData -> testGetSMADData_ifSuccess()

I'll add small description soon, thanks for point me that.
I found a reason why bad error message occurs when current or future date was chosen, problem solved (commit pushed)
In future Register and Login links will work properly.
Tests to be refactored, thanks for info

File: src/main/java/com/cebix/investmenttrackerapp/services/StockService.java [hunk 4]
```diff
@@ -28,11 +29,17 @@ public Stock retrieveStockData(String ticker, String date) {
```
Comment (Arturnix) [inline src/main/java/com/cebix/investmenttrackerapp/services/StockService.java:29]: This method _public Stock retrieveStockData(String ticker, String date)_ has few hardcoded operations which are performed inside it. It maybe a good approach to put some block of codes to separate private methods.

For example here You can make checker method which takes _ticker_ as an parameter.

File: src/main/java/com/cebix/investmenttrackerapp/services/StockService.java [hunk 5]
```diff
@@ -28,11 +29,17 @@ public Stock retrieveStockData(String ticker, String date) {
 
         LocalDate selectedDate;
```
Comment (Arturnix) [inline src/main/java/com/cebix/investmenttrackerapp/services/StockService.java:30]: This hardcoded method could be moved to separate private method to provide selectedDate as an output of this method.

File: src/main/java/com/cebix/investmenttrackerapp/services/StockService.java (hunk 4 above)
Comment (Cebix90) [reply src/main/java/com/cebix/investmenttrackerapp/services/StockService.java:27]: Good idea, extracted some parts of code to smaller private methods

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 6]
```diff
@@ -63,52 +65,56 @@ public void deleteUser(String customUserEmail) {
...
-            throw new IllegalArgumentException(e.getMessage(), e);
         }
     }
 
-    private void validateEmail(String email) {
-        if (email == null || email.trim().isEmpty()) {
-            throw new IllegalArgumentException("Email must not be null or empty");
-        }
+    private boolean validateEmail(String email) {
+        return email != null && !email.trim().isEmpty();
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:99]: This does not check the email format

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java [hunk 7]
```diff
@@ -63,52 +65,56 @@ public void deleteUser(String customUserEmail) {
...
     }
 
-    private void validatePassword(String password) {
-        if (password == null || password.trim().isEmpty()) {
-            throw new IllegalArgumentException("Password must not be null or empty");
-        }
+    private boolean validatePassword(String password) {
+        return password != null && !password.trim().isEmpty();
```
Comment (blazejprzyluski) [inline src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:103]: Password could use more validation

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java [hunk 8]
```diff
@@ -26,11 +30,13 @@ public class CustomUserDAOTests {
                 .withReuse(true);
 
     CustomUserDAO customUserDAO;
+    PortfolioDAO portfolioDAO;
```
Comment (blazejprzyluski) [inline src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java:33]: private?

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java [hunk 9]
```diff
@@ -174,42 +180,42 @@ public void testUpdateUserPassword_whenNewEmailIsEmpty_thenThrowsException() {
...
-            assertThrows(IllegalArgumentException.class, () -> customUserDAO.updateUserPortfolio("test@example.com", ""));
-        }
     }
 
     private CustomUser createUserForTests(String email) {
         CustomUser newUser = new CustomUser();
         newUser.setEmail(email);
         newUser.setPassword("password123");
-        newUser.setPortfolio("Test Portfolio");
+        newUser.setPortfolio(null);
```
Comment (blazejprzyluski) [inline src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java:218]: why null?

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAOTests.java [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+    @Container
+    PostgreSQLContainer postgresqlContainer = (PostgreSQLContainer) new PostgreSQLContainer(postgres)
+            .withDatabaseName("test_investment_tracker")
+            .withUsername("test")
+            .withPassword("test")
+            .withReuse(true);
+
+    CustomUserDAO customUserDAO;
+    PortfolioDAO portfolioDAO;
+    CustomUser userForTests;
```
Comment (blazejprzyluski) [inline src/test/java/com/cebix/investmenttrackerapp/databaseutils/PortfolioDAOTests.java:33]: private

File: src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java (hunk 9 above)
Comment (Cebix90) [reply src/test/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAOTests.java:218]: user can be created without Portfolio, during CustomUserDao tests portfolio is not necessary

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java (hunk 6 above)
Comment (Cebix90) [reply src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:99]: added the email format check

File: src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java (hunk 7 above)
Comment (Cebix90) [reply src/main/java/com/cebix/investmenttrackerapp/databaseutils/CustomUserDAO.java:103]: Good idea, I added restriction for password

**Pull request merged**